### PR TITLE
OAS 3.0 schema generation

### LINKED ISSUE #604: Support DRF 3.10+ generateschema (OAS 3)
DRF 3.10 is adding support to generate an openapi 3.0 schema: https://github.com/encode/django-rest-framework/pull/6532. There will likely be some enhancements needed for DJA to support this new functionality.

### LINKED ISSUE #644: missing DjangoFilterBackend.get_schema_operation_parameters()
Upstream DRF's `generateschema` requires filter backends to have a `get_schema_operation_parameters()` in order to generate a query parameter schema. This is missing from DjangoFilterBackend. This is a pre-req for #604.

### PR DESCRIPTION
**Pull request by keyz182**
Fixes #604
Fixes #644 

## Description of the Change

Extends DRF >= 3.12's generateschema to produce a jsonapi-formatted OAS schema document.
Builds on @n2ygk's work in https://github.com/django-json-api/django-rest-framework-json-api/pull/689
- Updated to latest DJA version

## Checklist

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] `CHANGELOG.md` updated (only for user relevant changes)
- [x] author name in `AUTHORS`

### CODE REVIEW

**Comment by keyz182** [comment]
There are some TODOs left from @n2ygk's work that I'm unsure of, and/or too unfamiliar to understand. However, for initial support for OAS 3.0, it is working well.

**Comment by n2ygk** [comment]
@keyz182 I will try to take a look at this later this week. Meanwhile, I note there's [ongoing activity w.r.t. openapi generateschema happening upstream at DRF](https://github.com/encode/django-rest-framework/pulls?utf8=%E2%9C%93&q=label%3A%22Schema+Generation%22+) including a bit of [flaming](https://github.com/encode/django-rest-framework/issues/7205) :-)  so you may want to build against DRF-master. (I haven't looked at your work yet, so maybe you are already doing this.)

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [inline rest_framework_json_api/schemas/openapi.py]: basically a copy? When I look at the [upsteram implementation](https://github.com/encode/django-rest-framework/blame/4a98533746db44c997882ba01d5515a29a61dcc3/rest_framework/schemas/openapi.py#L141) it seems that it looks quite different. Could it be that this has heavily changed upstream since DJA work has been started on OAS?

We need to be careful copying code as it becomes a burden maintaining such code. It would be good to double check how this code can be best implemented and bring closer how DRF does it. And if overwriting the whole method is really necessary better to adjust this comment why overwriting is actually needed.

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [inline rest_framework_json_api/schemas/openapi.py]: Is this still the case that this needs to be overwritten? I don't quite see why this should be different in DJA then in DRF.

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [inline rest_framework_json_api/schemas/openapi.py]: upstream there is a method `_get_serializer` which does this error handling.

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [inline rest_framework_json_api/schemas/openapi.py]: there is a `self._get_serializer` method.

**Review by sliverc — DISMISSED** [review]
Thanks for picking this up. It will be great to have OAS in DJA. All in all is this PR still very big. I have tried to go through it but I am sure I have overlooked things. So far I haven't looked at the tests yet but just the implementation because of lack of time. Once my comments are addressed I will need to review again anyway.

All in all when you look through the PR again it is important that you could look at:
* what code blocks are really necessary for basic OAS support (one comment I have added concerning this but there might be more). All what can be moved to a later PR makes this PR be approved quicker.
*  update with upstream code. It seems upstream has moved quite a bit and this PR might be simplified therefore (some comments I have added concerning this).

**Comment by keyz182** [comment]
Thanks for the feedback @sliverc - I'll try to get back and address your comments later this week (I have some other priorities to start this week off).

**Comment by n2ygk** [comment]
FYI [DRF 3.12 milestone](https://github.com/encode/django-rest-framework/milestone/76) continues to develop OAS [schema definition](https://github.com/encode/django-rest-framework/issues?q=is%3Aopen+is%3Aissue+label%3A%22Schema+Generation%22)

**Comment by keyz182** [comment]
Sorry - given current events, and new priorities at work, I'm not going to be able to address your concerns just now @sliverc. I will get back to it though.

**Comment by auvipy** [comment]
> Sorry - given current events, and new priorities at work, I'm not going to be able to address your concerns just now @sliverc. I will get back to it though.

hey! can you pick this up again?

**Comment by auvipy** [comment]
I will try to handle it if no one got the time. will try to split the PR in smaller chunks

**Comment by keyz182** [comment]
> > Sorry - given current events, and new priorities at work, I'm not going to be able to address your concerns just now @sliverc. I will get back to it though.
> 
> hey! can you pick this up again?

@auvipy Sorry, I'm not going to have time for a few months at least unfortunately :(

If you want to take over, I'm happy to give pointers where I can.

**Comment by auvipy** [comment]
OK I will take this over and feel free to review.

File: example/serializers.py
Comment (auvipy) [reply example/serializers.py:36]: yup these changes can be reverted

File: example/tests/test_serializers.py
Comment (auvipy) [reply example/tests/test_serializers.py:10]: I will revert this in my PR

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:466]: DRF was coming up with non-unique operation_ids and I think it still does that.

Also note that drf has renamed many of the private methods to public (dropping the leading `_`) so there's some hope of stability:-)

**Comment by n2ygk** [comment]
I've just had a quick look at current drf master (still pre-3.12) and have noticed a few things that might impact this code if not already dealt with:
- many private methods are now public (remove the _). The private versions are still there but will raise a warning.
- DRF now implements `components` and has a few other changes in `get_schema()` but it looks like it still has to be fully overridden because of the need to "expand" endpoints for `relationship` and `related` views. An idea for a future PR to DRF might be to add a method that can be more cleanly extended.

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:455]: yes, upstream changed a lot. He inlined something that had been a function call.

**Comment by n2ygk** [comment]
FYI - I've got some free time this week so started looking at this, especially w.r.t. changes to upstream DRF and will be pushing a rebase shortly.

**Comment by auvipy** [comment]
> FYI - I've got some free time this week so started looking at this, especially w.r.t. changes to upstream DRF and will be pushing a rebase shortly.

Thats great! Looking forwarded to it

File: requirements/requirements-optionals.txt
Comment (n2ygk) [reply requirements/requirements-optionals.txt:3]: @sliverc I'm not sure how to do this. Can you provide some guidance?

Comment (n2ygk) [reply requirements/requirements-optionals.txt:3]: oh, I see.... something like this:
```
    extras_require={
        'django-polymorphic': ['django-polymorphic>=2.0'],
        'django-filter': ['django-filter>=2.0'],
        'openapi': ['pyyaml>=5.3', 'uritemplate>=3.0.1']
    },
```

**Comment by n2ygk** [comment]
working my way backwards from master through earlier versions of DRF to eventually get tox tests to conditionalize as appropriate and otherwise run cleanly.

**Comment by n2ygk** [comment]
See also PR #804 which I've copied the change over here just to keep this moving and cleanly rebaseable.

**Comment by sliverc** [comment]
I won't have the time right now to closely look at this as in DRF there has been many changes and this is a fairly large PR.

@n2ygk I guess you won't see this as ready though right? As there are quite a few TODOs in the code. And possibly as there are many changes in DRF we should properly also wait till 3.12 is released.

**Comment by n2ygk** [comment]
No, still a draft. I have to check out a few more items and fix them to
align with DRF 3.12.

**Comment by n2ygk** [comment]
This is getting closer to reviewable. I'll do another rebase when I'm ready and will then request your review.

**Comment by n2ygk** [comment]
Just rebasing to stay current. Still a few items to get to.

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py]: DRF is still coming up with non-unique operation_ids due to the multiple paths for a given view. For example:
```yaml
/v1/courses/{id}/:
/v1/courses/{id}/relationships/course_terms/:
/v1/courses/{id}/course_terms/:
```
So this overrides by using the path.

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:615]: I've fixed this to use upstream's `get_serializer`

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py]: comments update to reflect the fact this is not a copy, although, unfortunately, the current structure of upstream DRF still requires full replacements of several methods rather than simply extending them.

**Comment by n2ygk** [comment]
@sliverc what is your opinion: Do I make this conditional on DRF [3.12](https://github.com/encode/django-rest-framework/milestone/76) (not yet released) rather then 3.11 (latest)? I'm starting to have to kludge things up to make it work with 3.11 given significant changes made in 3.12. It might best just be to hold this until 3.12 comes out. Anyone who wants it early can grab DRF master and this branch.

What do you think?

**Comment by auvipy** [comment]
> @sliverc what is your opinion: Do I make this conditional on DRF [3.12](https://github.com/encode/django-rest-framework/milestone/76) (not yet released) rather then 3.11 (latest)? I'm starting to have to kludge things up to make it work with 3.11 given significant changes made in 3.12. It might best just be to hold this until 3.12 comes out. Anyone who wants it early can grab DRF master and this branch.
> 
> What do you think?

it's apparent that drf 3.12 will be released on september. IMHO it would be best served with DRF 3.12/master.

**Comment by n2ygk** [comment]
> it's apparent that drf 3.12 will be released on september. IMHO it would be best served with DRF 3.12/master.

@auvipy I'm curious how you concluded that?

**Comment by auvipy** [comment]
> > it's apparent that drf 3.12 will be released on september. IMHO it would be best served with DRF 3.12/master.
> 
> @auvipy I'm curious how you concluded that?

I follow DRF issues & commits & it's core maintainers activities. they intended to release 3.12, 3-4 months earlier but later got stuck for many reasons. now that the author cleared most of his backlogs on other projects he & other members are willing to put more effort on 3.12 release. that is how I concluded O:) that might be pretty optimistic though :-1:

**Comment by sliverc** [comment]
> @sliverc what is your opinion: Do I make this conditional on DRF [3.12](https://github.com/encode/django-rest-framework/milestone/76) (not yet released) rather then 3.11 (latest)? I'm starting to have to kludge things up to make it work with 3.11 given significant changes made in 3.12. It might best just be to hold this until 3.12 comes out. Anyone who wants it early can grab DRF master and this branch.
> 
> What do you think?

Yes best wait till DRF 3.12 is released. Once this is done we can always look at whether this additional feature only supports DRF >3.12 or whether we might drop support of lower DRF versions for whole of DJA as well.
But as said let's wait for the release and if you can keep this PR up-to-date. That would be great.

**Comment by n2ygk** [comment]
Will keep it up to date. I have one last review comment of yours to address
and then I intend to re-request your review.

**Review by n2ygk — APPROVED** [review]

**Comment by sliverc** [comment]
Once this is updated with master again I will do another review.

**Comment by n2ygk** [comment]
@sliverc OK, ready for your re-review. One of my PRs didn't make it into DRF 3.12 so the last commit addresses that just to pass coveralls. If and when it is included I can add some more tests and remove the `pragma: no cover`

File: example/tests/test_openapi.py
Comment (sliverc) [inline example/tests/test_openapi.py:107]: by only reading the comment is this still needed as we are now only supporting DRF 3.12+?

File: example/tests/test_openapi.py
Comment (sliverc) [inline example/tests/test_openapi.py:139]: all tests above are pytest style. To be consistent better to convert this to pytest as well.

**Review by sliverc — CHANGES_REQUESTED** [review]
:sweat_smile: finally made it through this PR.

@n2ygk @keyz182 
When addressing the raised issues can I ask you NOT TO amend push but add simple commits for each change? As this is an enourmous PR it makes it easier for me to follow up. We will squash commit at the end anyway.

Thanks for your great work on this!

File: CHANGELOG.md
Comment (n2ygk) [reply CHANGELOG.md:24]: yeah I was just trying to clean things up. Will leave as is with two separate lines for 1.11 and 2.1.

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:389]: @sliverc I have to research this one some more. The declaration of this RelationShipView is dynamic, for example:
```
    url(r'^entries/(?P<pk>[^/.]+)/relationships/(?P<related_field>\w+)$',
        EntryRelationshipView.as_view(),
        name='entry-relationships'),
```

But those relationship URLs are produced by the EntryView so I have to find the way back to the EntryView given all I have from the EntryRelationshipView is the Entry model:
```
class EntryRelationshipView(RelationshipView):
    queryset = Entry.objects.all()
```

And this certainly is incorrect right now because the Entry schema generated looks like this (with 10 relationships):
```yaml
    Entry:
      type: object
      required:
      - type
      - id
      additionalProperties: false
      properties:
        type:
          $ref: '#/components/schemas/type'
        id:
          $ref: '#/components/schemas/id'
        links:
          type: object
          properties:
            self:
              $ref: '#/components/schemas/link'
        attributes:
          type: object
          properties:
            headline:
              type: string
              maxLength: 255
            body_text:
              type: string
              nullable: true
            pub_date:
              type: string
              format: date
              nullable: true
            mod_date:
              type: string
              format: date
              nullable: true
            body_format:
              type: string
              readOnly: true
          required:
          - headline
        relationships:
          type: object
          properties:
            blog:
              $ref: '#/components/schemas/reltoone'
            blog_hyperlinked:
              $ref: '#/components/schemas/reltoone'
            authors:
              $ref: '#/components/schemas/reltomany'
            comments:
              $ref: '#/components/schemas/reltomany'
            comments_hyperlinked:
              $ref: '#/components/schemas/reltomany'
            featured:
              $ref: '#/components/schemas/reltoone'
            suggested:
              $ref: '#/components/schemas/reltoone'
            suggested_hyperlinked:
              $ref: '#/components/schemas/reltoone'
            tags:
              $ref: '#/components/schemas/reltomany'
            featured_hyperlinked:
              $ref: '#/components/schemas/reltoone'
```

but the relationships paths only have those (3) from the Model:
```
/entries/{id}/relationships/authors
/entries/{id}/relationships/comments
/entries/{id}/relationships/tags
```

BTW, the command to generate the schema is:
`django-admin generateschema --generator_class rest_framework_json_api.schemas.openapi.SchemaGenerator --settings=example.settings`

Maybe the right approach to this is to make use of the EntryViewSet to generate the `relationships` paths and ignore the EntryRelationshipView.

I'll dig some more.

File: example/tests/test_openapi.py
Comment (n2ygk) [reply example/tests/test_openapi.py:139]: Yeah that was superfluous. I wonder why I did that?

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:389]: Maybe I've overthought this and can just put a templated URL in the schema:
```yaml
paths:
  /entries/{id}/relationships/{related_field}:
    get:
      operationId: List/entries/{id}/relationships/{related_field}
      description: ''
      parameters:
      - name: id
        in: path
        required: true
        description: A unique integer value identifying this entry.
        schema:
          type: string
      - name: related_field
        in: path
        required: true
        description: the name of the related field
        schema:
          type: string
```

There's no real reason to have to expand out all the possible related field names here.... And, this URL will likely only be used via HATEOAS link from the `/entries/{id}` response.

Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:389]: ... and a future enhancement would be to change the `related_field` to include an enumeration of allowed values (probably via $ref to components/parameters/Entry_related_fields):

```
      - name: related_field
        in: path
        required: true
        description: the name of the related field
        schema:
          type: string
          enum:
            - blog
            - blog_hyperlinked
            - authors
            - comments
            - comments_hyperlinked
            - featured
            - suggested
            - suggested_hyperlinked
            - tags
            - featured_hyperlinked
```

Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:389]: This makes a nicer swagger-ui:

Screen Shot 2020-10-12 at 11 29 08 AM

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:326]: It's in 
```
    def get_request_body(self, path, method, action=None):
        """
        A request body is required by jsonapi for POST, PATCH, and DELETE methods.
        This has an added parameter which is not in upstream DRF:

        :param action: None for conventional path; 'rel' or 'rels' for a singular or plural
            relationship of a related path, respectively.
        """
```

The action gets changed e.g. from `List` (for GET) to `rel` or `rels` for singular or plural relationships since the return schema is singular `"data": {...}`  vs. plural `"data": [ ... ]`

**Comment by n2ygk** [comment]
@sliverc back to you!

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py]: On further review I think I need to revert 5855b65 and revisit it since relationships can be either singular or plural and that depends on the individual relationship definitions.

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [reply rest_framework_json_api/schemas/openapi.py:319]: I think I have found the method which relies on a patched view which is `is_list_view`.

**Comment by sliverc** [comment]
@n2ygk went through the conversations and marked the once solved which are. Besides the relationship issue there are also some conversations though which you might have overlooked.
When you go to https://github.com/django-json-api/django-rest-framework-json-api/pull/772#pullrequestreview-503789975 there is a link "7 hidden conversations Load more..." which I assume you did not see?

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:490]: Hmm, it's a toss-up. I supposed I could change the base depending on the particular path (main/related/relationships) but is it necessary? Why not just use the path? The allowable character set for operation IDs is "string".

Comment (sliverc) [reply rest_framework_json_api/schemas/openapi.py:490]: I am not suggesting to change the logic but to overwrite less logic of DRF. At a quick glance it seems that simply overwriting `get_operation_id_base` like below instead of `get_operation_id` would lead to the same result as now. 

```python
def get_operation_id_base(self, path, method, action):
   return path
```

Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:490]: Oh, I see. I misread the recently-added description of get_operation_id_base. I thought it was a class attribute not a method. It was added in 3.12. I suppose I can do that. Will take a deeper look.

Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:490]: Actually operation_id_base is an Autoschema parameter - so that is the static part I was thinking of. get_operation_id_base is *much* longer than get_operation_id and the net result is I would have to do extra work to determine if the path is for the base collection/item or for the related or relationships.... Not sure it's worth doing that just to get a unique operationID that nobody even cares about but is required by the OAS spec.

I stared at this for about an hour and it's harder than it appears. There are larger unsolved issues I'd like to focus on for this initial release like fixing `_expand_relationships()`

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:617]: This one overrides an upstream method (and does in fact return a value).

Agree with the naming of the various private methods and will fix those.

File: rest_framework_json_api/schemas/openapi.py
Comment (n2ygk) [reply rest_framework_json_api/schemas/openapi.py:319]: I confused myself in the prior response.

To be more clear about what's happening here...  With  kwarg`{'get': 'retrieve_related'}` the path that starts at, e.g. AuthorViewSet, but returns the views for each `<related_field>` so I am using the related view's schema but meanwhile `self.view` is still the "parent" View (Authors in this example).
```
    url(r'^authors/(?P<pk>[^/.]+)/(?P<related_field>\w+)/$',
        AuthorViewSet.as_view({'get': 'retrieve_related'}),
        name='author-related'),
```

```yaml
/authors/{id}/comments/:
    # GET of the AuthorView results in the schema for CommentView.
```
And since the relationship can be toOne or toMany, the action is set to either `fetch` or `list` in to get the singular or plural response schema.

**Comment by n2ygk** [comment]
@sliverc please hold off on further review until I solve https://github.com/django-json-api/django-rest-framework-json-api/pull/772#discussion_r500961721

**Comment by sliverc** [comment]
@n2ygk Do you think you will get to the open OAS issues in the next few days?

If not once #846 is merged I would be fine if we release 4.0.0 nowish (mainly to support newest DRF version) and postpone this to the next minor version 4.1.0 as it is fully backwards compatible. What do you think?

**Comment by n2ygk** [comment]
Yes I think I have a solution for that problem I was working

**Comment by n2ygk** [comment]
@sliverc Ready for your review.

File: rest_framework_json_api/schemas/openapi.py
Comment (sliverc) [reply rest_framework_json_api/schemas/openapi.py]: I like how handling of `RelationshipView` has been simplified now.

**Review by sliverc — APPROVED** [review]

**Comment by sliverc** [comment]
Thanks for your work on this! This now looks good as a start and can be vetted as it is used more widely.

**Comment by keyz182** [comment]
Nice work @n2ygk !

**Comment by n2ygk** [comment]
@sliverc It looks like travis has stalled.

**Comment by n2ygk** [comment]
> @sliverc It looks like travis has stalled.

Not stalled but running really slowly. So far 3+ hours in and not done yet.

**Comment by n2ygk** [comment]
> > @sliverc It looks like travis has stalled.
> 
> Not stalled but running really slowly. So far 3+ hours in and not done yet.

Hmm, [this tweet](https://twitter.com/dwijnand/status/1321873017299939329?s=20) says something about .org vs. .com

**Pull request merged**
